File: seamstress/src/args.c
#include "args.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#define ARG_BUF_SIZE 64

struct args {
  char script_file[ARG_BUF_SIZE];
  char local_port[ARG_BUF_SIZE];
  char remote_port[ARG_BUF_SIZE];
  char width[ARG_BUF_SIZE];
  char height[ARG_BUF_SIZE];
};

static struct args a = {
  .script_file = "script",
  .local_port = "7777",
  .remote_port = "6666",
  .width = "256",
  .height = "128",
};

int args_parse(int argc, char **argv) {
  int opt;
  while ((opt = getopt(argc, argv, "x:y:s:l:b:?:h")) != -1) {
    switch (opt) {
    case 'x':
      strncpy(a.width, optarg, ARG_BUF_SIZE - 1);
      break;
    case 'y':
      strncpy(a.height, optarg, ARG_BUF_SIZE - 1);
      break;
    case 's':
      strncpy(a.script_file, optarg, ARG_BUF_SIZE - 1);
      break;
    case 'l':
      strncpy(a.local_port, optarg, ARG_BUF_SIZE - 1);
    case 'b':
      strncpy(a.remote_port, optarg, ARG_BUF_SIZE -1);
      break;
    case '?':
    case 'h':
    default:
      fprintf(stdout, "Start seamstress with optional overrides:\n");
      fprintf(stdout, "-s   override user script [default %s]\n", a.script_file);
      fprintf(stdout, "-l   override OSC listen port [default %s]\n", a.local_port);
      fprintf(stdout, "-b   override OSC broadcast port [default %s]\n", a.remote_port);
      fprintf(stdout, "-x   override window width [default %s]\n", a.width);
      fprintf(stdout, "-y   override window height [default %s]\n", a.height);
      exit(1);
      ;
    }
  }
  return 0;
}

const char *args_local_port(void) {
  return a.local_port;
}

const char *args_remote_port(void) {
  return a.remote_port;
}

const char *args_script_file(void) {
  return a.script_file;
}

int args_width(void) {
  return atoi(a.width);
}

int args_height(void) {
  return atoi(a.height);
}

File: src/args.zig
const std = @import("std");

pub var script_file: [:0]const u8 = "script";
pub var local_port: [:0]const u8 = "7777";
pub var remote_port: [:0]const u8 = "6666";
pub var width: [:0]const u8 = "256";
pub var height: [:0]const u8 = "128";

pub fn parse() !void {
    var args = std.process.args();
    var i: u8 = 0;
    while (args.next()) |*arg| : (i += 1) {
        if (i == 0) {
            continue;
        }
        if ((arg.len != 2) or (arg.*[0] != '-')) {
            break;
        }
        switch (arg.*[1]) {
            'b' => {
                if (args.next()) |next| {
                    remote_port = next;
                    continue;
                }
            },
            'l' => {
                if (args.next()) |next| {
                    local_port = next;
                    continue;
                }
            },
            's' => {
                if (args.next()) |next| {
                    script_file = next;
                    continue;
                }
            },
            'x' => {
                if (args.next()) |next| {
                    width = next;
                    continue;
                }
            },
            'y' => {
                if (args.next()) |next| {
                    height = next;
                    continue;
                }
            },
            else => {
                break;
            },
        }
        break;
    } else {
        args.deinit();
        return;
    }
    args.deinit();
    try print_usage();
    std.process.exit(1);
}

fn print_usage() !void {
    const stdout_file = std.io.getStdOut().writer();
    var bw = std.io.bufferedWriter(stdout_file);
    const stdout = bw.writer();
    try stdout.print("Start seamstress-zig with optional overrides:\n", .{});
    try stdout.print("-s    override user script [current {s}]\n", .{script_file});
    try stdout.print("-l    override OSC listen port [current {s}]\n", .{local_port});
    try stdout.print("-b    override OSC broadcast port [current {s}]\n", .{remote_port});
    try stdout.print("-x    override window width [current {s}]\n", .{width});
    try stdout.print("-y    override window height [current {s}]\n", .{height});
    try bw.flush();
}
